text/0000-extension-layer.md

```diff
@@ -0,0 +1,90 @@
+# Meta
+[meta]: #meta
+- Name: Add extension layer to exchange data
+- Start Date: 2023-10-09
+- Author(s): [c0d1ngm0nk3y](https://github.com/c0d1ngm0nk3y), [pbusko](https://github.com/pbusko)
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request:
+- CNB Pull Request:
+- CNB Issue:
+- Related: [RFC#0105 Support Dockerfiles](https://github.com/buildpacks/rfcs/blob/main/text/0105-dockerfiles.md)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This RFC introduces support for Extension Layers to allow data transfer between the build environment and the Kaniko execution.
+
+# Motivation
+[motivation]: #motivation
+
+This change allows extensions to possess their own layers to utilize during the generation/extend process. Additionally, it ensures that extension output does not inadvertently interfere with other extension or buildpack layers during the build, and it does not unintentionally become part of the final application image.
+
+This would allow distroless run images to be extended.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This follows up on RFC-0105 and proposes that during the execution of the extension's `bin/generate`, an extension is allowed to write arbitrary data to its exclusive layer. This data then becomes accessible during the execution of the `extend` phase via Kaniko context. The content of these extension-specific layers is ignored at build and launch time, it serves only the extension phase.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Before execution of the `bin/generate`, the lifecycle will create a distinct writable layer for each extension which passed detection. The extensions can then write to these layers, and the Kaniko context is set to the corresponding layer during the `extend` phase instead of the `<app>` directory. The Extension Layers will not be included in the final image by the lifecycle.
+
+The location of the Extension Layer will be provided to the `bin/generate` via additional environment variable `CNB_EXT_LAYER_DIR`.
+
+# Migration
+[migration]: #migration
+
+Since we would change the Kaniko context, this would be breaking for existing extensions.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+The workspace would no longer be the Kaniko context. But the benefit of having it in the first place seems quite limited anyway.
```

Comment:
Perhaps we could denote the location of the app directory with a `CNB_APP_DIR` env var

Reply:
Is the idea that by using the new ext-layer dir as the context we don't have to deal with volumes in the Dockerfiles?


```
FROM <blah>
# because we downloaded some-expected-file.sh into $CNB_OUTPUT/ext-layer
RUN some-expected-file.sh
```
If the extension needed something from the `/app` directory, would we expect extension authors to copy it to `/ext-layer` then?

Maybe `$CNB_OUTPUT/context` instead of `layer` would make it clearer?

Reply:
> Is the idea that by using the new ext-layer dir as the context we don't have to deal with volumes in the Dockerfiles?

Exactly.

> If the extension needed something from the /app directory, would we expect extension authors to copy it to /ext-layer then?

Since extensions are not allowed to write to the `/app` directory is there any valid reason to copy files from there? Given the fact that extensions are allowed to read, it should be perfectly fine to copy necessary stuff to the dedicated folder.